embedded/document/engine.go

```diff
@@ -322,34 +346,41 @@ func (e *Engine) DeleteCollection(ctx context.Context, collectionName string) er
 }
 
 func (e *Engine) InsertDocument(ctx context.Context, collectionName string, doc *structpb.Struct) (docID DocumentID, txID uint64, err error) {
-	docID, txID, _, err = e.upsertDocument(ctx, collectionName, doc, true)
+	opt := newUpsertOption().withIsInsert(true)
+	docID, txID, _, err = e.upsertDocument(ctx, collectionName, doc, opt)
 	return
 }
 
 func (e *Engine) UpdateDocument(ctx context.Context, collectionName string, doc *structpb.Struct) (txID, rev uint64, err error) {
-	_, txID, rev, err = e.upsertDocument(ctx, collectionName, doc, false)
+	opt := newUpsertOption().withIsInsert(false)
+	_, txID, rev, err = e.upsertDocument(ctx, collectionName, doc, opt)
 	return
 }
 
-func (e *Engine) upsertDocument(ctx context.Context, collectionName string, doc *structpb.Struct, isInsert bool) (docID DocumentID, txID, rev uint64, err error) {
+func (e *Engine) upsertDocument(ctx context.Context, collectionName string, doc *structpb.Struct, opt *upsertOption) (docID DocumentID, txID, rev uint64, err error) {
```

Comment:
based on previous implementation of upsert in SQL layer, just `isInsert` option may be needed, so dynamic options may be too much, but it they are added it may be with a dynamic number and all the options applied e.g. for _,opt:= range options { opt(someObj) } ... but as far as I can see there is no need for it...

Reply:
There is also a need to pass in an existing transaction, and then it leads to too many args in the function, hence added an option struct for it